typescript/infra/src/funding/adapters/EVMFundingAdapter.ts

```diff
@@ -0,0 +1,415 @@
+ * Extends BaseEvmAdapter to leverage EVM-specific functionality
+ */
+export class EVMFundingAdapter
+  extends BaseEvmAdapter
+  implements IFundingAdapter
+{
+  private readonly localMultiProvider: MultiProvider;
+  private readonly igp: HyperlaneIgp;
+  private readonly nativeBridges: Record<
+    string,
+    { l1ETHGateway: string; l1Messenger: string }
+  >;
+  private readonly L2Chains: ChainName[];
+  private readonly L2ToL1: Record<ChainName, ChainName>;
+
+  constructor(
+    chainName: ChainName,
+    private readonly multiProtocolProvider: MultiProtocolProvider,
+    environment: DeployEnvironment,
+    private readonly context: Contexts,
+    private readonly fundingAddresses: FundingAddresses,
+    public readonly logger: Logger,
+    private readonly walletBalanceGauge: Gauge<string>,
+  ) {
+    super(chainName, multiProtocolProvider, fundingAddresses, logger);
+    this.localMultiProvider = multiProtocolProvider.toMultiProvider();
+    this.localMultiProvider.setSigner(chainName, this.getSigner());
+    this.igp = HyperlaneIgp.fromAddressesMap(
+      {
+        [this.chainName]: this.fundingAddresses,
+      },
+      this.localMultiProvider,
+    );
+    this.nativeBridges = {
+      scrollsepolia: {
+        l1ETHGateway: '0x8A54A2347Da2562917304141ab67324615e9866d',
+        l1Messenger: '0x50c7d3e7f7c656493D1D76aaa1a836CedfCBB16A',
+      },
+    };
+    this.L2Chains = L2_CHAINS;
+    this.L2ToL1 = {
+      ...L2_CHAINS.reduce(
+        (acc, chain) => {
+          acc[chain] = L1_CHAIN;
+          return acc;
+        },
+        {} as Record<ChainName, ChainName>,
+      ),
+    };
+  }
+
+  async getBalance(address: string): Promise<string> {
+    return (await this.getProvider().getBalance(address)).toString();
+  }
+
+  async getFundingAmount(
+    address: string,
+    desiredBalance: number,
+    fundingThresholdFactor: number,
+    role: FundableRole,
+  ): Promise<string> {
+    const currentBalance = BigNumber.from(await this.getBalance(address));
+    const desiredBalanceBigNumber = ethers.utils.parseEther(
+      desiredBalance.toString(),
+    );
+
+    const delta = desiredBalanceBigNumber.sub(currentBalance);
+    // TODO: check if this is correct
+    const minDelta = desiredBalanceBigNumber
+      .mul(Math.floor(fundingThresholdFactor * 100))
+      .div(100);
+
+    this.logger.info(
+      {
+        chain: this.chainName,
+        currentBalance: ethers.utils.formatEther(currentBalance),
+        desiredBalance: ethers.utils.formatEther(desiredBalanceBigNumber),
+        delta: ethers.utils.formatEther(delta),
+        minDelta: ethers.utils.formatEther(minDelta),
+        fundingThresholdFactor,
+        role,
+      },
+      'Funding amount',
+    );
+    return delta.gt(minDelta) ? delta.toString() : '0';
+  }
+
+  async fundKey(
+    key: BaseAgentKey,
+    desiredBalance: number,
+    fundingThresholdFactor: number,
+  ): Promise<void> {
+    if (await this.shouldBridgeToL2()) {
```

Comment:
said already but just marking the l1->l2 bridging design decision again next to some more of the code we could consider removing